examples/deal.II/bps_02.cc

```diff
@@ -0,0 +1,465 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+//  Authors: Peter Munch, Martin Kronbichler
+//
+// ---------------------------------------------------------------------
+
+#include "bps.h"
+
+// deal.II includes
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/parameter_handler.h>
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/distributed/shared_tria.h>
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_renumbering.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+
+#include <deal.II/matrix_free/cuda_fe_evaluation.h>
+#include <deal.II/matrix_free/cuda_matrix_free.h>
+
+// boost
+#include <boost/algorithm/string.hpp>
+
+#include <sstream>
+
+// include operators
+
+// Test cases
+// TESTARGS(name="BP1") --resource {ceed_resource} --bp BP1 --fe_degree 2 --print_timings 0
+// TESTARGS(name="BP4") --resource {ceed_resource} --bp BP5 --fe_degree 3 --print_timings 0
+
+/**
+ * Relevant parameters.
+ */
+struct Parameters
+{
+  BPType       bp                   = BPType::BP5;
+  unsigned int n_global_refinements = 1;
+  unsigned int fe_degree            = 2;
+  bool         print_timings        = true;
+  std::string  libCEED_resource     = "/cpu/self/avx/blocked";
+
+  bool
+  parse(int argc, char *argv[])
+  {
+    if (argc == 1 && (std::string(argv[0]) == "--help"))
+      {
+        std::cout << "Usage: ./bp [OPTION]..." << std::endl;
+        std::cout << std::endl;
+        std::cout << "--bp             name of benchmark (BP1, BP5)" << std::endl;
+        std::cout << "--n_refinements  number of refinements (0-)" << std::endl;
+        std::cout << "--fe_degree      polynomial degree (1-)" << std::endl;
+        std::cout << "--print_timings  name of benchmark (0, 1)" << std::endl;
+        std::cout << "--resource       name of resource (e.g., /cpu/self/avx/blocked)" << std::endl;
```

Comment:
@jeremylt I am wondering whether it is possible to extract the resource information from Kokkos and pass it to libCEED or what the best approach is to keep them in sync?

Reply:
All libCEED backends can accept CPU pointers for vectors, so what we do in PETSc is

1) Create the user requested libCEED backend

2) Request the preferred memory space (CEED_MEM_HOST or CEED_MEM_DEVICE) of this libCEED backend

3) Set PETSc to use the requested memory space

I think a simpler version of this could be to request the libCEED backend's preferred memory space and hand off pointers in that memory space.

https://github.com/CEED/libCEED/blob/e4483ef5325cb386ca74ee375c1e72625464d2e6/examples/petsc/multigrid.c#L111